pyro/ops/streaming.py

```diff
@@ -0,0 +1,196 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import copy
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import Dict, Type, Union
+
+import torch
+
+from pyro.ops.welford import WelfordCovariance
+
+
+class StreamingStats(ABC):
+    """
+    Abstract base class for streamable statistics of trees of tensors.
+
+    Derived classes must implelement :meth:`update`, :meth:`merge`, and
+    :meth:`get`.
+    """
+
+    @abstractmethod
+    def update(self, sample) -> None:
+        """
+        Update state from a single sample.
+
+        This mutates ``self`` and returns nothing. Updates should be
+        independent of order, i.e. samples should be exchangeable.
+
+        :param sample: A sample value which is a nested dictionary of
+            :class:`torch.Tensor` leaves. This can have arbitrary nesting and
+            shape shape, but assumes shape is constant across calls to
+            ``.update()``.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def merge(self, other: "StreamingStats") -> "StreamingStats":
+        """
+        Select two aggregate statistics, e.g. from different MCMC chains.
+
+        This is a pure function: it returns a new :class:`StreamingStats`
+        object and does not modify either ``self`` or ``other``.
+
+        :param other: Another streaming stats instance of the same type.
+        """
+        assert isinstance(other, type(self))
+        raise NotImplementedError
+
+    @abstractmethod
+    def get(self) -> dict:
+        """
+        Return the aggregate statistic, which should be a dictionary.
+        """
+        raise NotImplementedError
+
+
+class CountStats(StreamingStats):
+    """
+    Statistic tracking only the number of samples.
+    """
+
+    def __init__(self):
+        self.count = 0
+        super().__init__()
+
+    def update(self, sample):
+        self.count += 1
+
+    def merge(self, other: "CountStats"):
+        assert isinstance(other, type(self))
```

Comment:
nit: these assertions should no longer be necessary with type hints

Reply:
Hmm, good point more generally.  However I'd like to argue that we should include both type hints and assertions until all common tools can leverage type hints.  My reasoning is that I'd really like to catch errors as early as possible, e.g. when users (like me) are working in a jupyter notebook.  I think until Jupyter dynamically checks types while editing we'll want extra guard rails especially for tricky interfaces like this.